File: Server/src/admin/adminService.ts
import bcrypt from 'bcryptjs'
import crypto from 'node:crypto'
import type { Pool } from 'pg'
import type { UserRole } from '../lib/contracts.js'

type CreateUserInput = {
  username: string
  password: string
  role: UserRole
  displayName?: string
}

type UpdateUserInput = {
  email?: string
  role?: UserRole
  displayName?: string | null
  isActive?: boolean
}

type DbUserRow = {
  id: string
  email: string
  username: string
  display_name: string | null
  role: UserRole
  is_active: boolean
  created_at: string
  updated_at: string
}

export class AdminServiceError extends Error {
  constructor(
    message: string,
    readonly statusCode = 400,
  ) {
    super(message)
    this.name = 'AdminServiceError'
  }
}

export class AdminService {
  constructor(private readonly pool: Pool) {}

  private buildManagedUserEmail(username: string) {
    return `${username.trim().toLowerCase()}@derton.local`
  }

  async getOverview() {
    const [users, sessions, instruments, ticks, watchlists] = await Promise.all([
      this.pool.query<{
        total_users: string
        active_users: string
        admin_users: string
      }>(
        `
          select
            count(*)::text as total_users,
            count(*) filter (where is_active = true)::text as active_users,
            count(*) filter (where role = 'admin' and is_active = true)::text as admin_users
          from users
        `,
      ),
      this.pool.query<{ active_sessions: string }>('select count(*)::text as active_sessions from sessions where expires_at > now()'),
      this.pool.query<{ instrument_count: string }>('select count(*)::text as instrument_count from instruments'),
      this.pool.query<{
        tick_count: string
        oldest_tick_at: string | null
        newest_tick_at: string | null
      }>(
        `
          select
            count(*)::text as tick_count,
            min(recorded_at)::text as oldest_tick_at,
            max(recorded_at)::text as newest_tick_at
          from market_ticks
        `,
      ),
      this.pool.query<{ watchlist_count: string }>('select count(*)::text as watchlist_count from watchlists'),
    ])

    const userRow = users.rows[0]
    const sessionRow = sessions.rows[0]
    const instrumentRow = instruments.rows[0]
    const tickRow = ticks.rows[0]
    const watchlistRow = watchlists.rows[0]

    return {
      users: {
        total: Number(userRow?.total_users ?? 0),
        active: Number(userRow?.active_users ?? 0),
        admins: Number(userRow?.admin_users ?? 0),
      },
      sessions: {
        active: Number(sessionRow?.active_sessions ?? 0),
      },
      instruments: {
        total: Number(instrumentRow?.instrument_count ?? 0),
      },
      watchlists: {
        total: Number(watchlistRow?.watchlist_count ?? 0),
      },
      marketHistory: {
        tickCount: Number(tickRow?.tick_count ?? 0),
        oldestTickAt: tickRow?.oldest_tick_at ?? null,
        newestTickAt: tickRow?.newest_tick_at ?? null,
      },
    }
  }

  async listUsers() {
    const result = await this.pool.query<
      DbUserRow & {
        active_session_count: string
        last_session_at: string | null
      }
    >(
      `
        select
          users.id,
          users.email,
          users.username,
          users.display_name,
          users.role,
          users.is_active,
          users.created_at::text,
          users.updated_at::text,
          count(sessions.id) filter (where sessions.expires_at > now())::text as active_session_count,
          max(sessions.created_at)::text as last_session_at
        from users
        left join sessions on sessions.user_id = users.id
        group by users.id
        order by users.created_at asc
      `,
    )

    return result.rows.map((row) => ({
      id: row.id,
      email: row.email,
      username: row.username,
      displayName: row.display_name,
      role: row.role,
      isActive: row.is_active,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
      activeSessionCount: Number(row.active_session_count ?? 0),
      lastSessionAt: row.last_session_at,
    }))
  }

  async createUser(actorUserId: string, input: CreateUserInput) {
    const username = input.username.trim().toUpperCase()
    const email = this.buildManagedUserEmail(username)
    const displayName = input.displayName?.trim() || null

    const existing = await this.pool.query(
      `
        select 1
        from users
        where lower(email) = lower($1) or upper(username) = upper($2)
        limit 1
      `,
      [email, username],
    )

    if (existing.rowCount) {
      throw new AdminServiceError('A user with this user ID already exists.', 409)
    }

    const id = crypto.randomUUID()
    const passwordHash = await bcrypt.hash(input.password, 10)

    await this.pool.query(
      `
        insert into users (id, email, username, display_name, password_hash, role, is_active, updated_at)
        values ($1, $2, $3, $4, $5, $6, true, now())
      `,
      [id, email, username, displayName, passwordHash, input.role],
    )

    await this.writeAudit(actorUserId, 'user.create', 'user', id, {
      email,
      username,
      role: input.role,
      displayName,
    })

    return id
  }

  async updateUser(actorUserId: string, userId: string, input: UpdateUserInput) {
    const user = await this.getUserRow(userId)
    if (!user) {
      throw new AdminServiceError('User not found.', 404)
    }

    const nextRole = input.role ?? user.role
    const nextIsActive = input.isActive ?? user.is_active

    if ((user.role === 'admin' && nextRole !== 'admin') || (user.role === 'admin' && nextIsActive === false)) {
      await this.ensureAnotherActiveAdminExists(user.id)
    }

    if (input.email) {
      const emailConflict = await this.pool.query(
        `
          select 1
          from users
          where lower(email) = lower($1) and id <> $2
          limit 1
        `,
        [input.email.trim().toLowerCase(), userId],
      )

      if (emailConflict.rowCount) {
        throw new AdminServiceError('Another user already uses this email.', 409)
      }
    }

    const updates: string[] = []
    const values: Array<string | boolean | null> = []
    let index = 1

    if (input.email !== undefined) {
      updates.push(`email = $${index++}`)
      values.push(input.email.trim().toLowerCase())
    }

    if (input.role !== undefined) {
      updates.push(`role = $${index++}`)
      values.push(input.role)
    }

    if (input.displayName !== undefined) {
      updates.push(`display_name = $${index++}`)
      values.push(input.displayName ? input.displayName.trim() : null)
    }

    if (input.isActive !== undefined) {
      updates.push(`is_active = $${index++}`)
      values.push(input.isActive)
    }

    updates.push(`updated_at = now()`)
    values.push(userId)

    await this.pool.query(
      `
        update users
        set ${updates.join(', ')}
        where id = $${index}
      `,
      values,
    )

    if (input.isActive === false) {
      await this.pool.query('delete from sessions where user_id = $1', [userId])
    }

    await this.writeAudit(actorUserId, 'user.update', 'user', userId, input)
  }

  async resetPassword(actorUserId: string, userId: string, password: string) {
    const user = await this.getUserRow(userId)
    if (!user) {
      throw new AdminServiceError('User not found.', 404)
    }

    const passwordHash = await bcrypt.hash(password, 10)
    await this.pool.query(
      `
        update users
        set password_hash = $2, updated_at = now()
        where id = $1
      `,
      [userId, passwordHash],
    )

    await this.pool.query('delete from sessions where user_id = $1', [userId])

    await this.writeAudit(actorUserId, 'user.reset_password', 'user', userId, {
      revokeSessions: true,
    })
  }

  async revokeSessions(actorUserId: string, userId: string) {
    const user = await this.getUserRow(userId)
    if (!user) {
      throw new AdminServiceError('User not found.', 404)
    }

    await this.pool.query('delete from sessions where user_id = $1', [userId])

    await this.writeAudit(actorUserId, 'user.revoke_sessions', 'user', userId, {})
  }

  private async getUserRow(userId: string) {
    const result = await this.pool.query<DbUserRow>(
      `
        select id, email, username, display_name, role, is_active, created_at::text, updated_at::text
        from users
        where id = $1
        limit 1
      `,
      [userId],
    )

    return result.rows[0] ?? null
  }

  private async ensureAnotherActiveAdminExists(excludedUserId: string) {
    const result = await this.pool.query<{ count: string }>(
      `
        select count(*)::text as count
        from users
        where role = 'admin'
          and is_active = true
          and id <> $1
      `,
      [excludedUserId],
    )

    if (Number(result.rows[0]?.count ?? 0) < 1) {
      throw new AdminServiceError('You cannot remove or deactivate the last active admin.', 409)
    }
  }

  private async writeAudit(
    actorUserId: string,
    action: string,
    entityType: string,
    entityId: string | null,
    payload: Record<string, unknown>,
  ) {
    await this.pool.query(
      `
        insert into audit_logs (id, actor_user_id, action, entity_type, entity_id, payload)
        values ($1, $2, $3, $4, $5, $6::jsonb)
      `,
      [crypto.randomUUID(), actorUserId, action, entityType, entityId, JSON.stringify(payload)],
    )
  }
}

File: Server/app/services/admin_service.py
from __future__ import annotations

import secrets

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.clickhouse import get_ch_client
from app.config import settings
from app.core.exceptions import ConflictError, NotFoundError
from app.core.security import hash_password, make_id


class AdminService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def get_overview(self) -> dict:
        users = await self.db.execute(
            text(
                """
                SELECT
                    COUNT(*)::text as total,
                    COUNT(*) FILTER (WHERE is_active = true)::text as active,
                    COUNT(*) FILTER (WHERE role = 'admin' AND is_active = true)::text as admins
                FROM users
                """
            )
        )
        u = users.one_or_none()
        sessions = await self.db.execute(
            text("SELECT COUNT(*)::text FROM sessions WHERE expires_at > now()")
        )
        instruments = await self.db.execute(
            text("SELECT COUNT(*)::text FROM instruments")
        )
        watchlists = await self.db.execute(
            text("SELECT COUNT(*)::text FROM watchlists")
        )

        try:
            ch = get_ch_client()
            tick_rows = ch.execute(
                "SELECT COUNT(*), MIN(recorded_at), MAX(recorded_at) FROM market_ticks"
            )
            t = tick_rows[0] if tick_rows else (0, None, None)
        except Exception:
            t = (0, None, None)

        return {
            "users": {
                "total": int(u[0] or 0),
                "active": int(u[1] or 0),
                "admins": int(u[2] or 0),
            },
            "sessions": {"active": int(sessions.scalar() or 0)},
            "instruments": {"total": int(instruments.scalar() or 0)},
            "watchlists": {"total": int(watchlists.scalar() or 0)},
            "market_history": {
                "tick_count": int(t[0] or 0),
                "oldest_tick_at": t[1].isoformat() if t[1] and hasattr(t[1], "isoformat") else t[1],
                "newest_tick_at": t[2].isoformat() if t[2] and hasattr(t[2], "isoformat") else t[2],
            },
        }

    async def list_users(self) -> list[dict]:
        result = await self.db.execute(
            text(
                """
                SELECT
                    users.id, users.email, users.username, users.display_name,
                    users.role, users.is_active, users.created_at::text, users.updated_at::text,
                    COUNT(sessions.id) FILTER (WHERE sessions.expires_at > now())::text as active_sessions,
                    MAX(sessions.created_at)::text as last_session
                FROM users
                LEFT JOIN sessions ON sessions.user_id = users.id
                GROUP BY users.id
                ORDER BY users.created_at ASC
                """
            )
        )
        return [
            {
                "id": r[0],
                "email": r[1],
                "username": r[2],
                "display_name": r[3],
                "role": r[4],
                "is_active": r[5],
                "created_at": r[6],
                "updated_at": r[7],
                "active_session_count": int(r[8] or 0),
                "last_session_at": r[9],
            }
            for r in result.all()
        ]

    async def create_user(self, actor_id: str, input_data: dict) -> str:
        username = input_data["username"].strip().upper()
        email = f"{username.lower()}@derton.local"
        display_name = input_data.get("display_name", "").strip() or None

        existing = await self.db.execute(
            text(
                """
                SELECT 1 FROM users
                WHERE LOWER(email) = LOWER(:email) OR UPPER(username) = UPPER(:username)
                LIMIT 1
                """
            ),
            {"email": email, "username": username},
        )
        if existing.one_or_none():
            raise ConflictError("A user with this username already exists.")

        uid = secrets.token_hex(16)
        pwd_hash = hash_password(input_data["password"])
        await self.db.execute(
            text(
                """
                INSERT INTO users (id, email, username, display_name, password_hash, role, is_active, updated_at)
                VALUES (:id, :email, :username, :display_name, :pwd_hash, :role, true, now())
                """
            ),
            {
                "id": uid,
                "email": email,
                "username": username,
                "display_name": display_name,
                "pwd_hash": pwd_hash,
                "role": input_data["role"],
            },
        )

        await self._write_audit(actor_id, "user.create", "user", uid, {
            "email": email, "username": username, "role": input_data["role"],
        })
        return uid

    async def update_user(self, actor_id: str, user_id: str, input_data: dict) -> None:
        user = await self._get_user(user_id)
        if not user:
            raise NotFoundError("User not found.")

        updates: list[str] = []
        params: dict = {"id": user_id}
        idx = 1

        if "email" in input_data and input_data["email"] is not None:
            updates.append(f"email = :p{idx}")
            params[f"p{idx}"] = input_data["email"].lower().strip()
            idx += 1
        if "role" in input_data and input_data["role"] is not None:
            updates.append(f"role = :p{idx}")
            params[f"p{idx}"] = input_data["role"]
            idx += 1
        if "display_name" in input_data:
            updates.append(f"display_name = :p{idx}")
            params[f"p{idx}"] = input_data["display_name"].strip() if input_data["display_name"] else None
            idx += 1
        if "is_active" in input_data and input_data["is_active"] is not None:
            updates.append(f"is_active = :p{idx}")
            params[f"p{idx}"] = input_data["is_active"]
            idx += 1

        if not updates:
            return

        updates.append("updated_at = now()")
        await self.db.execute(
            text(f"UPDATE users SET {', '.join(updates)} WHERE id = :id"),
            params,
        )

        if input_data.get("is_active") is False:
            await self.db.execute(
                text("DELETE FROM sessions WHERE user_id = :uid"), {"uid": user_id}
            )

        await self._write_audit(actor_id, "user.update", "user", user_id, input_data)

    async def reset_password(self, actor_id: str, user_id: str, password: str) -> None:
        user = await self._get_user(user_id)
        if not user:
            raise NotFoundError("User not found.")

        pwd_hash = hash_password(password)
        await self.db.execute(
            text("UPDATE users SET password_hash = :pwd, updated_at = now() WHERE id = :id"),
            {"pwd": pwd_hash, "id": user_id},
        )
        await self.db.execute(
            text("DELETE FROM sessions WHERE user_id = :uid"), {"uid": user_id}
        )
        await self._write_audit(actor_id, "user.reset_password", "user", user_id, {})

    async def revoke_sessions(self, actor_id: str, user_id: str) -> None:
        user = await self._get_user(user_id)
        if not user:
            raise NotFoundError("User not found.")
        await self.db.execute(
            text("DELETE FROM sessions WHERE user_id = :uid"), {"uid": user_id}
        )
        await self._write_audit(actor_id, "user.revoke_sessions", "user", user_id, {})

    async def _get_user(self, user_id: str) -> dict | None:
        result = await self.db.execute(
            text(
                """
                SELECT id, email, username, display_name, role, is_active
                FROM users WHERE id = :id LIMIT 1
                """
            ),
            {"id": user_id},
        )
        row = result.one_or_none()
        if not row:
            return None
        return {
            "id": row[0], "email": row[1], "username": row[2],
            "display_name": row[3], "role": row[4], "is_active": row[5],
        }

    async def _write_audit(
        self, actor_id: str, action: str, entity_type: str, entity_id: str | None, payload: dict
    ) -> None:
        await self.db.execute(
            text(
                """
                INSERT INTO audit_logs (id, actor_user_id, action, entity_type, entity_id, payload)
                VALUES (:id, :actor, :action, :entity_type, :entity_id, :payload::jsonb)
                """
            ),
            {
                "id": secrets.token_hex(16),
                "actor": actor_id,
                "action": action,
                "entity_type": entity_type,
                "entity_id": entity_id,
                "payload": payload,
            },
        )
